File: app/src/main/java/com/mutenlab/sudoit/ui/MainActivity.java
package com.mutenlab.sudoit.ui;

import android.Manifest;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;

import com.mutenlab.sudoit.R;
import com.mutenlab.sudoit.util.PermissionHelper;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * @author icerrate
 */
public class MainActivity extends AppCompatActivity {

    private final static int REQUEST_CAMERA_PERMISSION = 100;

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        toolbar.setTitle(R.string.app_name);
        setSupportActionBar(toolbar);
    }

    @OnClick(R.id.start_camera)
    public void startCameraClick(View view) {
        String[] permission = new String[]{Manifest.permission.CAMERA};

        if (!PermissionHelper.checkPermissions(this, permission)) {
            ActivityCompat.requestPermissions(this, permission, REQUEST_CAMERA_PERMISSION);
        } else {
            startActivity(new Intent(this, CameraActivity.class));
        }
    }
}

File: app/src/main/kotlin/com.mutenlab.sudoit/MainActivity.kt
package com.mutenlab.sudoit

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.widget.Button
import android.widget.Toast
import com.mutenlab.sudoit.ui.CameraActivity
import com.mutenlab.sudoit.utils.PermissionHelper

class MainActivity : AppCompatActivity() {

    companion object {
        const val REQUEST_MEDIA_PERMISSIONS = 100
        private val TAG = MainActivity::class.qualifiedName
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val btnStartOcr = findViewById<Button>(R.id.start_ocr)
        btnStartOcr.setOnClickListener {
            checkCameraPermission()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            REQUEST_MEDIA_PERMISSIONS -> if (grantResults.size == 2
                    && grantResults[0] == PackageManager.PERMISSION_GRANTED
                    && grantResults[1] == PackageManager.PERMISSION_GRANTED) {
                launchCamera()
            } else {
                Toast.makeText(this, getString(R.string.camera_request_rationale),
                        Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun checkCameraPermission() {
        val mediaPermissions = arrayOf(Manifest.permission.CAMERA,
                Manifest.permission.WRITE_EXTERNAL_STORAGE)
        if (!PermissionHelper.checkPermissions(this, mediaPermissions)) {
            Log.i(MainActivity.TAG, "Permission to media denied")
            ActivityCompat.requestPermissions(this, mediaPermissions, REQUEST_MEDIA_PERMISSIONS)
        } else {
            launchCamera()
        }
    }

    private fun launchCamera() {
        val intent = Intent(applicationContext, CameraActivity::class.java)
        startActivity(intent)
    }
}
